cpp/src/arrow/memory_pool.h

```diff
@@ -175,6 +177,31 @@ ARROW_EXPORT Status jemalloc_memory_pool(MemoryPool** out);
 ARROW_EXPORT
 Status jemalloc_set_decay_ms(int ms);
 
+/// \brief Get basic statistics from jemalloc's mallctl.
+/// See the MALLCTL NAMESPACE section in jemalloc project documentation for
+/// available stats.
+ARROW_EXPORT
+Result<uint64_t> jemalloc_get_stat(const char* name);
+
+/// \brief Reset the counter for peak bytes allocated in the calling thread to zero.
+/// This affects subsequent calls to thread.peak.read, but not the values returned by
+/// thread.allocated or thread.deallocated.
+ARROW_EXPORT
+Status jemalloc_peak_reset();
+
+/// \brief Print summary statistics in human-readable form to stderr.
+/// See malloc_stats_print documentation in jemalloc project documentation for
+/// available opt flags.
+ARROW_EXPORT
+Status jemalloc_stats_print(std::function<void(void*, const char*)>* write_cb,
+                            void* cbopaque, const char* opts = "");
```

Comment:
There is no need to pass a `void*` as C++ conveniently allows to pass a closure (or any callable object) as a `std::function`.

Reply:
(of course you must arrange for the `std::function` to be kept alive, however)

Reply:
So this should probably be:
```c++
Status jemalloc_stats_print(std::function<void(const char*)> write_cb, const char* opts = "");
```
(or perhaps keep passing `write_cb` as a pointer, which feels less idiomatic but can avoid a move/copy)

Reply:
I can't figure out a way to pass `std::function<void(void*, const char*)>*` to `malloc_stats_print` without getting `signal 11: SIGSEGV`.
As per [jemalloc docs](https://jemalloc.net/jemalloc.3.html): "The malloc_stats_print() function writes summary statistics via the write_cb callback function pointer and cbopaque data passed to write_cb, or malloc_message() if write_cb is NULL."
It also states: "The malloc_message variable allows the programmer to override the function which emits the text strings forming the errors and warnings if for some reason the STDERR_FILENO file descriptor is not suitable for this. malloc_message() takes the cbopaque pointer argument that is NULL unless overridden by the arguments in a call to malloc_stats_print(), followed by a string pointer. Please note that doing anything which tries to allocate memory in this function is likely to result in a crash or deadlock."

Any idea how to proceed?

Reply:
I'll take a look in the next days...

Reply:
Actually, @benibus , would you like to help @rok on this? (or even suggest a patch that he can apply :-))

Reply:
Yep! I'll take a look at it.

Reply:
I should have suggestions up later today.